larq/quantizers.py

```diff
@@ -601,9 +604,95 @@ def get_config(self):
         return {**super().get_config(), "k_bit": self.precision}
 
 
+@utils.register_alias("dorefa_kernel_quantizer")
+@utils.register_keras_custom_object
+class DoReFaKernel(DoReFa):
+    r"""Instantiates a serializable k_bit kernel quantizer as in the DoReFa paper.
+
+    This quantizer is the same like the `DoReFa` quantizer, but adds a preprocessing.
+    Instead of limiting input operands (or in this case: weights) using a hard
+    limiter, a tangens hyperbolicus is applied to achieve a softer limiting
+    with a gradient, which is contineously differentiable itself.
+
+    \\[
+    w_{lim}(w) = tanh(w)
+    \\]
+
+    Furthermore, the weights of each layer are normed, such that the weight with
+    the largest magnitude gets the largest or smallest (depending on its sign)
+    quantizable value. That way, the full quantizable numeric range is utilized.
+
+    \\[
+    w_{norm}(w) = \frac{w}{max(|w|)}
+    \\]
+
+    The formulas can be found in the paper in section 2.3. Please note, that
+    the paper refers to weights being quantized on a numeric range of [-1,1], while
+    activations are quantized on the numeric range [0,1]. `DoReFa` defines the
+    quantization function quantizek() from the paper with the correct numeric
+    range of [0,1], which is why the range needs to be adapted before applying
+    `DoReFa` after applying the preprocessing (limit and norm).
+    The hard limiting inside `DoReFa` becomes ineffective, because its input
+    is already limited by the hyperbolic tangent. The full quantization
+    function including the adaption of numeric ranges is
+
+    \\[
+    q(w) = 2quantize_{k}(\frac{w_{norm}\left(w_{lim}\left(w\right)\right)}{2} + \frac{1}{2}) - 1
+    \\]
+
+    !!! warning
+        This quantizer works for weights on the range [-1,1], which matches the
+        default setting of `constraints.weight_clip`. Do not use this quantizer
+        with a different constraint *clip_value* than the default one.
+
+    ```plot-activation
+    quantizers.DoReFaKernel
+    ```
+
+    The interface of this quantizer is the same like the one of `DoReFa`.
+
+    # References
+        - [DoReFa-Net: Training Low Bitwidth Convolutional Neural Networks with Low
+            Bitwidth Gradients](https://arxiv.org/abs/1606.06160)
+    """
+
+    def call(self, inputs):
+
+        limited = tf.math.tanh(inputs)
+
+        # Divider for max-value norm.
+        # If all elements are 0., we would get a div by zero.
+        # So when all elements are zero, nothing is normed. This is achieved by
+        # dividing by 1.
```

Comment:
Why is this necessary? I don't see any mention of this in the paper and the linked [reference implementation](https://github.com/tensorpack/tensorpack/blob/5845ead80e3d77c89e2f1e9d5e0691324a8354cc/examples/DoReFa-Net/dorefa.py#L35) also doesn't seem to do this.

TensorFlow will return `inf` for  `x / 0`, which obviously isn't ideal, but I don't think this code would throw a divide by zero error.

Reply:
If this would ocur in one layer, it would output _inf_ for at least one output pixel. The following batch norm layer would calculate the mean of some usual numbers and one _inf_, which is _inf_. The BN layer would apply a _- inf_ operation to all pixels to normalize the mean and the _inf_ value spreads to all pixels. So if this occurs in one spot, it ruins the entire network. Having all weights at exactly zero is anyhow a rare occasion, but having _inf_ or _-inf_ for all outputs is probably not what the user expects. So I added this check. Do you agree?

Reply:
Makes sense. I looked at the code again and the only way this could happen is when all element of `limited` are exactly zero. This means we can replace `limited / dividend` with [`tf.math.divide_no_nan(limited, dividend)`](https://www.tensorflow.org/api_docs/python/tf/math/divide_no_nan) which returns `0` in cases `dividend == 0` which allows to remove the special handling on line 667 without changing the computation.

Reply:
Done in b4a6e29db3a1a6caaa0503d8896395badc14b659